cmd/broker/fanout/main.go

```diff
@@ -0,0 +1,200 @@
+	log.Printf("Registering %d informers", len(injection.Default.GetInformers()))
+
+	ctx, informers := injection.Default.SetupInformers(ctx, cfg)
+
+	var env envConfig
+	if err := envconfig.Process("", &env); err != nil {
+		log.Fatal("Failed to process env var", zap.Error(err))
+	}
+
+	kubeClient := kubeclient.Get(ctx)
+
+	// We sometimes startup faster than we can reach kube-api. Poll on failure to prevent us terminating
+	if perr := wait.PollImmediate(time.Second, 60*time.Second, func() (bool, error) {
+		if err = version.CheckMinimumVersion(kubeClient.Discovery()); err != nil {
+			log.Printf("Failed to get k8s version %v", err)
+		}
+		return err == nil, nil
+	}); perr != nil {
+		log.Fatal("Timed out attempting to get k8s version: ", err)
+	}
+
+	// Set up our logger.
+	loggingConfig, err := sharedmain.GetLoggingConfig(ctx)
+	if err != nil {
+		log.Fatal("Error loading/parsing logging configuration: ", err)
+	}
+
+	logger, atomicLevel := pkglogging.NewLoggerFromConfig(loggingConfig, component)
+	defer flush(logger)
+	ctx = pkglogging.WithLogger(ctx, logger)
+
+	// Run informers instead of starting them from the factory to prevent the sync hanging because of empty handler.
+	if err := controller.StartInformers(ctx.Done(), informers...); err != nil {
+		logger.Fatalw("Failed to start informers", zap.Error(err))
+	}
+
+	configMapWatcher := configmap.NewInformedWatcher(kubeClient, system.Namespace())
+	// Watch the logging config map and dynamically update logging levels.
+	configMapWatcher.Watch(pkglogging.ConfigMapName(), pkglogging.UpdateLevelFromConfigMap(logger, atomicLevel, component))
+	// Watch the observability config map
+	configMapWatcher.Watch(metrics.ConfigMapName(), metrics.UpdateExporterFromConfigMap(component, logger))
+
+	// Give the signal channel some buffer so that reconciling handlers won't
+	// block the targets config update?
+	targetsUpdateCh := make(chan struct{})
+	targetsConifg, err := volume.NewTargetsFromFile(
+		volume.WithPath(env.TargetsConifgPath),
+		volume.WithNotifyChan(targetsUpdateCh))
+	if err != nil {
+		logger.Fatalw("Failed to load targets config", zap.String("path", env.TargetsConifgPath), zap.Error(err))
+	}
+
+	logger.Info("Starting the broker fanout")
+
+	syncSignal := poolSyncSignal(ctx, targetsUpdateCh)
+	if _, err := fanout.StartSyncPool(ctx, targetsConifg, buildPoolOptions(env, syncSignal)...); err != nil {
+		logger.Fatalw("Failed to start fanout sync pool", zap.Error(err))
+	}
+
+	// Context will be done if a TERM signal is issued.
+	<-ctx.Done()
+	// Wait a grace period for the handlers to shutdown.
+	time.Sleep(30 * time.Second)
+	logger.Info("Done waiting, exit.")
+}
+
+func poolSyncSignal(ctx context.Context, targetsUpdateCh chan struct{}) chan struct{} {
+	// Give it some buffer so that multiple signal could queue up
+	// but not blocking the signaler?
+	ch := make(chan struct{}, 10)
+	ticker := time.NewTicker(15 * time.Second)
```

Comment:
Yes. The 15s is like the resync period in case if the last self-reconcile wasn't completed fully successful.

Reply:
I wrote similar code for #739. It might not apply here but I thought I'd mention it.

I use an unbuffered channel for the sync signal and do a nonblocking send, since if there's already something in the channel, it means the update is happening now and will happen again immediately (I think): https://github.com/google/knative-gcp/pull/739/files#diff-99d1987d2919a2cce497502d2ecee2b0R498-R505

I also moved the ticker inside the channel reader goroutine so it's just another clause of the select: https://github.com/google/knative-gcp/pull/739/files#diff-99d1987d2919a2cce497502d2ecee2b0R481-R495

Reply:
Thanks for sharing. I think it might be better to fire the signal in a goroutine so that I can drop the channel buffer. But I didn't get your 2nd point. My code looks similar as yours?

Reply:
In #739 the select clause reading from the ticker executes the code synchronously rather than enqueueing to another channel. Seems like that doesn't apply to your case since you have multiple readers of the returned syncSignal channel (I think).

Reply:
Not by multiple reader. But my reader only accepts a single channel for signal. Here I’m trying to merge multiple signals.